interceptor/tracing_interceptor.go

```diff
@@ -0,0 +1,476 @@
+		Name:      info.WorkflowType.Name,
+		Tags: map[string]string{
+			workflowIDTagKey: info.WorkflowExecution.ID,
+			runIDTagKey:      info.WorkflowExecution.RunID,
+		},
+		FromHeader: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	var finishOpts TracerFinishSpanOptions
+	defer span.Finish(&finishOpts)
+
+	ret, err := t.Next.ExecuteWorkflow(ctx, in)
+	finishOpts.Error = err
+	return ret, err
+}
+
+type tracingWorkflowOutboundInterceptor struct {
+	WorkflowOutboundInterceptorBase
+	tracer Tracer
+}
+
+func (t *tracingWorkflowOutboundInterceptor) ExecuteActivity(
+	ctx workflow.Context,
+	activityType string,
+	args ...interface{},
+) workflow.Future {
+	// Start span writing to header
+	span, ctx, err := t.startNonReplaySpan(ctx, "StartActivity", activityType, true)
+	if err != nil {
+		return err
+	}
+	defer span.Finish(&TracerFinishSpanOptions{})
+
+	return t.Next.ExecuteActivity(ctx, activityType, args...)
+}
+
+func (t *tracingWorkflowOutboundInterceptor) ExecuteLocalActivity(
+	ctx workflow.Context,
+	activityType string,
+	args ...interface{},
+) workflow.Future {
+	// Start span writing to header
+	span, ctx, err := t.startNonReplaySpan(ctx, "StartActivity", activityType, true)
+	if err != nil {
+		return err
+	}
+	defer span.Finish(&TracerFinishSpanOptions{})
+
+	return t.Next.ExecuteLocalActivity(ctx, activityType, args...)
+}
+
+func (t *tracingWorkflowOutboundInterceptor) ExecuteChildWorkflow(
+	ctx workflow.Context,
+	childWorkflowType string,
+	args ...interface{},
+) workflow.ChildWorkflowFuture {
+	// Start span writing to header
+	span, ctx, err := t.startNonReplaySpan(ctx, "StartChildWorkflow", childWorkflowType, false)
+	if err != nil {
+		return err
+	}
+	defer span.Finish(&TracerFinishSpanOptions{})
+
+	return t.Next.ExecuteChildWorkflow(ctx, childWorkflowType, args...)
+}
+
+func (t *tracingWorkflowOutboundInterceptor) NewContinueAsNewError(
+	ctx workflow.Context,
+	wfn interface{},
+	args ...interface{},
+) error {
+	err := t.Next.NewContinueAsNewError(ctx, wfn, args...)
+	if !workflow.IsReplaying(ctx) {
+		if contErr, _ := err.(*workflow.ContinueAsNewError); contErr != nil {
+			// Get the current span and write header
+			if span, _ := ctx.Value(t.tracer.SpanContextKey()).(TracerSpan); span != nil {
+				if writeErr := writeSpanToHeader(t.tracer, span, WorkflowHeader(ctx)); writeErr != nil {
+					return fmt.Errorf("failed writing span when creating continue as new error: %w", writeErr)
+				}
+			}
+		}
+	}
+	return err
+}
+
+type nopSpan struct{}
+
+func (nopSpan) Finish(*TracerFinishSpanOptions) {}
+
+// Span always returned, even in replay. futErr is non-nil on error.
+func (t *tracingWorkflowOutboundInterceptor) startNonReplaySpan(
+	ctx workflow.Context,
+	operation string,
+	name string,
+	dependedOn bool,
+) (span TracerSpan, newCtx workflow.Context, futErr workflow.ChildWorkflowFuture) {
+	// Noop span if replaying
+	if workflow.IsReplaying(ctx) {
+		return nopSpan{}, ctx, nil
+	}
```

Comment:
It's not clear to me how much this will no-op out but I don't think we want to squelch all spans ever when replaying. You lose a lot of info that can be helpful when debugging. At minimum you probably want a way to at least turn these on

Reply:
This replicates the existing behavior in Go and the behavior at least as seen in Java. Workflow outbound spans are just not created during replay. Is this something we want to make configurable in all APIs?

Reply:
OK if that's how it is already then I'm fine leaving it that way. This works a little differently in core but all the spans are already different there anyway so no biggie.